genesis/applier.go

```diff
@@ -1,27 +1,144 @@
 package genesis
 
 import (
+	"context"
+	_ "embed"
 	"encoding/json"
 	"fmt"
 	"math/big"
+	"strings"
+	"sync"
 
-	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum-optimism/optimism/op-service/predeploys"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+//go:embed generated/l2-allocs/901-l2-allocs.json
+var l2AllocsJSON []byte
+
+type alloc struct {
+	Balance string            `json:"balance"`
+	Code    string            `json:"code"`
+	Nonce   string            `json:"nonce"`
+	Storage map[string]string `json:"storage"`
+}
+
+type allocs map[string]alloc
+
+type predeploy struct {
+	proxy common.Address
+	impl  common.Address
+}
+
+var interopProxyPredeploys = []common.Address{
+	predeploys.CrossL2InboxAddr,
+	predeploys.L2toL2CrossDomainMessengerAddr,
+	predeploys.L1BlockAddr,
+}
+
+const (
+	emptyCode = "0x"
 )
 
-func ApplyChainID(genesisJson []byte, chainId *big.Int) ([]byte, error) {
-	var genesis core.Genesis
-	if err := json.Unmarshal(genesisJson, &genesis); err != nil {
-		return nil, fmt.Errorf("unable to parse genesis json: %w", err)
+// Uses same logic as `predeployToCodeNamespace` in Predeploy.sol to determine impl address.
+func predeployToCodeNamespace(addr common.Address) common.Address {
+	addrBigInt := new(big.Int).SetBytes(addr.Bytes())
+
+	namespaceBigInt := new(big.Int).SetBytes(common.FromHex("0xc0D3C0d3C0d3C0D3c0d3C0d3c0D3C0d3c0d30000"))
+	resultBigInt := new(big.Int).Or(new(big.Int).And(addrBigInt, big.NewInt(0xffff)), namespaceBigInt)
+
+	return common.BytesToAddress(resultBigInt.Bytes())
+}
+
+func fetchAllocForAddr(addr common.Address, allocsJSON []byte) (*alloc, error) {
+	var allocs allocs
+	err := json.Unmarshal(allocsJSON, &allocs)
```

Comment:
we can prob memoize the result of the unmarshal and store in memory so we don't have to unmarshal everytime. esp since this is called at the startup we should try to reduce latency as much possible

Reply:
good call, memoized it at the top level call `opsimulator.configureInteropForChain`